backend/src/user/user.service.ts

```diff
@@ -171,4 +219,89 @@ export class UserService {
     });
     return res;
   }
+
+  /**
+   * Sends the user a reset password email and returns the JWT token
+   * @param email the email account that requested for a password reset
+   */
+
+  async resetPassword(email: string) {
+    // find user via email provided
+    const [err, user] = await this.findByEmail(email);
+    const username = user.username;
+
+    // if user doesn't exist, do nothing. Just bring user to email sent page
+    if (err || !user) {
+      return;
+    }
+
+    // check if email in cache, print out that the reset mail has already been sent
+    const isEmailInCache = !!(await this.cache.get(
+      RESET_PASSWORD_EMAIL_PREFIX + email
+    ));
+    if (isEmailInCache) {
+      throw new ForbiddenException(AUTH_ERROR.RESET_EMAIL_ALREADY_SENT);
+    }
+
+    //Reset password
+    const resetPasswordVerificationToken = RESET_PASSWORD_PREFIX + v4();
+    await this.cache.set<CacheableResetEmail>(resetPasswordVerificationToken, {
+      email,
+    });
+
+    //Generate a random temporary password for user of length 16
+    const temporaryPassword = randomUUID().slice(0,16);
+
+    const [error, userToReset] = await this.updateResetPassword(
+      user.id,
+      temporaryPassword
+    );
+
+    //Send email
+    await this.mailerService
+      .sendMail({
+        to: email,
+        subject: "Email Verification for resetting of password ✔",
+        template: "resetPasswordVerification", // The `.pug` or `.hbs` extension is appended automatically.
+        context: {
+          // Data to be sent to template engine.
+          code: resetPasswordVerificationToken,
+          username: username,
+          newPassword: temporaryPassword,
+        },
+      })
+      .then((success) => {
+        console.log(success);
+      })
+      .catch((err) => {
+        console.log(err);
+      });
+  }
+
+  async verifyResetEmail(token: string, confirmationPassword: string, newPassword: string) {
+    const cachedUser = await this.cache.get<CacheableResetEmail>(token);
+    if (!cachedUser) {
+      throw new ForbiddenException(AUTH_ERROR.INVALID_EMAIL_VERIFY_EMAIL_TOKEN);
+    }
+
+    const { email } = cachedUser;
+    const [err, user] = await this.findByEmail(email, true);
+    const userId = user.id;
+    const currentPassword = user.hash;
+
+    //Check whether password entered matches the password in the email
+    const isPasswordCorrect = await argon2.verify(currentPassword, confirmationPassword);
+
+    if (!isPasswordCorrect) {
+      throw new ForbiddenException(AUTH_ERROR.INVALID_CREDENTIALS);
+    }
+    
+    const [error, userToReset] = await this.updatePasswordWithHash(
+      userId,
+      newPassword
+    );
```

Comment:
remember to handle this error (that could occur if updating the password failed)